### ISSUE TITLE
MainActivity missing from classes.dex

### ISSUE THREAD

**Issue by Nutomic** [clojure-android/lein-droid#105]
```
E/AndroidRuntime( 6416): java.lang.RuntimeException: Unable to instantiate activity ComponentInfo{jaaka.core/jaaka.core.MainActivity}: java.lang.ClassNotFoundException: Didn't find class "jaaka.core.MainActivity" on path: DexPathList[[zip file "/data/app/jaaka.core-2.apk"],nativeLibraryDirectories=[/data/app-lib/jaaka.core-2, /vendor/lib, /system/lib]]
E/AndroidRuntime( 6416):        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2110)
E/AndroidRuntime( 6416):        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2233)
E/AndroidRuntime( 6416):        at android.app.ActivityThread.access$800(ActivityThread.java:135)
E/AndroidRuntime( 6416):        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
E/AndroidRuntime( 6416):        at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime( 6416):        at android.os.Looper.loop(Looper.java:136)
E/AndroidRuntime( 6416):        at android.app.ActivityThread.main(ActivityThread.java:5001)
E/AndroidRuntime( 6416):        at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 6416):        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
E/AndroidRuntime( 6416):        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
E/AndroidRuntime( 6416): Caused by: java.lang.ClassNotFoundException: Didn't find class "jaaka.core.MainActivity" on path: DexPathList[[zip file "/data/app/jaaka.core-2.apk"],nativeLibraryDirectories=[/data/app-lib/jaaka.core-2, /vendor/lib, /system/lib]]
E/AndroidRuntime( 6416):        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
E/AndroidRuntime( 6416):        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
E/AndroidRuntime( 6416):        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
E/AndroidRuntime( 6416):        at android.app.Instrumentation.newActivity(Instrumentation.java:1061)
E/AndroidRuntime( 6416):        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2101)
E/AndroidRuntime( 6416):        ... 9 more
E/AndroidRuntime( 6416):        Suppressed: java.lang.ClassNotFoundException: jaaka.core.MainActivity
E/AndroidRuntime( 6416):                at java.lang.Class.classForName(Native Method)
E/AndroidRuntime( 6416):                at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
E/AndroidRuntime( 6416):                at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
E/AndroidRuntime( 6416):                at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
E/AndroidRuntime( 6416):                ... 12 more
E/AndroidRuntime( 6416):        Caused by: java.lang.NoClassDefFoundError: Class "Ljaaka/core/MainActivity;" not found
E/AndroidRuntime( 6416):                ... 16 more
W/ActivityManager(  837):   Force finishing activity jaaka.core/.MainActivity
```

This seems to happen quite randomly, without changing any code. `dexdump target/classes.dex | grep "MainActivity"` finds nothing.

`lein clean; rm -r assets/ gen/` does not fix it either, I have to do clean, then `git stash; lein droid doall; git stash pop`. Maybe something is cached somewhere?

I can also send you the project files, but it's basically `lein droid new` with minor changes.

Btw, I feel like there should be a `lein droid clean` call that also deletes `gen/` and `assets/` directories ;)

Edit: lein-droid is version 0.2.3, `$ lein -v` output: `Leiningen 2.5.0 on Java 1.7.0_65 OpenJDK 64-Bit Server VM`, on Kubuntu 14.04.

**Comment by alexander-yakushev** [comment]
Since there was no more input into this, I close the issue as non-reproducible. Feel free to reopen it if the problem persists.

**Comment by wiseman** [comment]
I've run into this same problem several times.  For a while I was doing `lein droid clean` before every build, but that was somewhat superstitious.  I'll see if I can reproduce it reliably.

**Comment by wiseman** [comment]
Just a few more details:

I'm using `[lein-droid "0.3.3"]`, `[org.clojure-android/clojure "1.7.0-alpha3" :use-resources true]`.

In a working build I see my main namespace get compiled:
```
[...]
Compiling manifold.deferred
Compiling com.lemondronor.ad1.main
Reflection warning, com/lemondronor/ad1/main.clj:87:3 - call to method speak can't be resolved (target class is unknown).
Compiling clojure.test
[...]
```
If I then check out the new code that doesn't work and do `lein droid doall`, I don't even see my main namespace get compiled:

```
Compiling manifold.deferred
Compiling clojure.test
```

At this point an APK containing my old code, not the new code, gets installed and run (successfully).  If I now do `lein droid clean && lein droid doall`, again I don't see the main namespace get compiled, and the app fails with the `java.lang.ClassNotFoundException: Didn't find class "com.lemondronor.ad1.MainActivity" on path: DexPathList...` exception.

**Comment by alexander-yakushev** [comment]
Wait, do you do `lein droid clean` or `lein clean`? Because the latter one is correct.

**Comment by alexander-yakushev** [comment]
Also, I would suggest upgrading to lein-droid `0.3.5` just be on the same page.

**Comment by alexander-yakushev** [comment]
Gotta close this again as non-reproducible.

**Comment by satchit8** [comment]
Encountered something similar.

Reproduction steps:

1. Check out foreclojure source.

2. Edit the ns require portion of main.clj to contain:

     [user :as user]) ; missing a closing square bracket

   instead of:

     [user :as user]])

3. Build apk, install, and run via:

     lein droid doall

The build should succeed but (at least here) main.clj is never compiled.

**Comment by satchit8** [comment]
* When main.clj file has correct ns macro use, output of:

     `lein droid doall`

   contains a line like:

     > Compiling <something>.main

* When main.clj has certain problematic ns macro usage, the output of the lein invocation above does not include such a line.

* compile-clojure (from lein-droid's compile.clj) contains the following code:

```clojure
      (let [nses (namespaces-to-compile project)
                        .
                        .
                        .
                  (doseq [namespace# '~nses]
                  (println "Compiling" namespace#)
                  (clojure.core/compile namespace#))
```

* Since the "Compiling <something>.main" line is missing from the output, it appears that nses is missing something corresponding to main.clj. 

* namespaces-to-compile calls bultitude/namespaces-on-classpath and there is the following text in bultitude's README:

 > By default Bultitude will ignore files that have unreadable ns forms in
 > them in order not to cause problems with projects that include moustache
 > templates on their classpath. Most functions take an ignore-unreadable?
 > arg which you can set to false to make it propagate exceptions from the
 > reader.

* The namespaces-to-compile call in question did not specify the ignore-unreadable? argument.

* The attached patch employs ignore-unreadable? with a value of false and leads to better behavior here.

[bultitude-nicety.txt](https://github.com/clojure-android/lein-droid/files/275552/bultitude-nicety.txt)

**Comment by alexander-yakushev** [comment]
@satchit8 This sums it up really well! Thanks for the patch, I committed it and published in `lein-droid 0.4.4`.
